test/semaphore/SemaphoreProxyTest.js

```diff
@@ -0,0 +1,126 @@
+    });
+
+    after(function () {
+        client.shutdown();
+        return Controller.shutdownCluster(cluster.id);
+    });
+
+    it('inits semaphore with 10 permits', function () {
+        return semaphore.init(10)
+            .then(function (res) {
+                expect(res).to.equal(true);
+                return semaphore.availablePermits()
+            })
+            .then(function (res) {
+                expect(res).to.equal(10);
+            });
+    });
+
+    it('acquires 10 permits ', function () {
+        return semaphore.init(10)
+            .then(function (res) {
+                expect(res).to.equal(true);
+                return semaphore.acquire(10);
+            })
+            .then(function () {
+                return semaphore.availablePermits();
+            })
+            .then(function (res) {
+                expect(res).to.equal(0);
+            });
+    });
+
+    it('drain', function () {
+        return semaphore.init(10)
+            .then(function (res) {
+                expect(res).to.equal(true);
+                return semaphore.drainPermits(10);
+            })
+            .then(function (res) {
+                expect(res).to.equal(10);
+                return semaphore.availablePermits();
+            })
+            .then(function (res) {
+                expect(res).to.equal(0);
+            });
+    });
+
+    it('reduces permits', function () {
+        return semaphore.init(10)
+            .then(function (res) {
+                expect(res).to.equal(true);
+                return semaphore.reducePermits(10);
+            })
+            .then(function () {
+                return semaphore.availablePermits();
+            })
+            .then(function (res) {
+                expect(res).to.equal(0);
+            });
+    });
+
+    it('releases 5 permits', function () {
+        return semaphore.init(10)
+            .then(function (res) {
+                expect(res).to.equal(true);
+                return semaphore.acquire(10);
+            })
+            .then(function () {
+                return semaphore.availablePermits();
+            })
+            .then(function (res) {
+                expect(res).to.equal(0);
+                return semaphore.release(5)
+            })
+            .then(function () {
+                return semaphore.availablePermits();
+            })
+            .then(function (res) {
+                expect(res).to.equal(5);
+            });
+    });
+
+    it('try acquire', function () {
+        return semaphore.init(10)
+            .then(function (res) {
+                expect(res).to.equal(true);
+                return semaphore.tryAcquire(10);
+            })
+            .then(function (res) {
+                expect(res).to.equal(true);
+                return semaphore.availablePermits();
+            })
+            .then(function (res) {
+                expect(res).to.equal(0);
+            });
+    });
```

Comment:
can you add tests for negative scenarios, too? For example we want make sure that one of the 2 acquire calls does not succeed if the semaphore has only 1 permit. Another scenario I want to test goes like this: Start 2 clients. Initiate a sempahore with 1 permit. Try to acquire the semaphore and make sure that only one of them proceeds. Also, in another test I would like to see the other client is able to proceed after the first client releases a permit on a semaphore both clients race to acquire. There needs to be a handful of these tests. The actual use case of this proxy is this.